array.h

```diff
@@ -254,41 +296,50 @@ template <index_t Extent>
 using fixed_interval = interval<dynamic, Extent>;
 
 /** Make an interval from a half-open range `[begin, end)`. */
-inline interval<> range(index_t begin, index_t end) {
+NDARRAY_INLINE NDARRAY_DECORATE_IF_CUDA
+interval<> range(index_t begin, index_t end) {
   return interval<>(begin, end - begin);
 }
-inline interval<> r(index_t begin, index_t end) {
+NDARRAY_INLINE NDARRAY_DECORATE_IF_CUDA
+interval<> r(index_t begin, index_t end) {
   return interval<>(begin, end - begin);
 }
 
 /** Make an interval from a half-open range `[begin, begin + Extent)`. */
 template <index_t Extent>
+NDARRAY_DECORATE_IF_CUDA
 fixed_interval<Extent> range(index_t begin) {
   return fixed_interval<Extent>(begin);
 }
 template <index_t Extent>
+NDARRAY_DECORATE_IF_CUDA
 fixed_interval<Extent> r(index_t begin) {
   return fixed_interval<Extent>(begin);
 }
 
 /** Placeholder object representing an interval that indicates keep
  * the whole dimension when used in an indexing expression. */
+// TODO(jiawen): Should this be constexpr?
```

Comment:
It wouldn't hurt. Since all of the template parameters are non-dynamic, this doesn't actually initialize anything, but it would enable use in templates (if that ever was needed...).

Reply:
I need to refresh my memory on constexpr variables. And then check in what memory space they're place in device code.

Reply:
./array.h:321:27: error: constexpr variable cannot have non-literal type 'const interval<0, -1>'
constexpr interval<0, -1> all, _;
                          ^
./array.h:187:7: note: 'interval<0, -1>' is not literal because it is not an aggregate and has no constexpr constructors other than copy or move constructors

Shucks. Will have to look into this later. Reverted.